### PR TITLE
fix: set key_metadata to Null by default

### LINKED ISSUE #753: Write `null` for `key_metadata` instead of empty bytes
See https://github.com/apache/iceberg-rust/pull/742#issuecomment-2515564242 for details.

### PR DESCRIPTION
**Pull request by feniljain**
## Issue Resolved

Closes #753  

## About

- Converted Manifest spec's `key_metadata` field to be `Option<Vec<u8>` instead of just `Vec<u8>`
- Updated tests to reflect the same
- Ran both integration tests and iceberg crate tests, all are passing
- Also ran #742 to make sure avro file is written correctly and it does fix the underlying issue

### CODE REVIEW

File: crates/iceberg/src/spec/manifest.rs [hunk 1]
```diff
@@ -1134,7 +1134,7 @@ impl DataFile {
         &self.upper_bounds
     }
     /// Get the Implementation-specific key metadata for the data file.
-    pub fn key_metadata(&self) -> &[u8] {
+    pub fn key_metadata(&self) -> &Option<Vec<u8>> {
```
Comment (Xuanwo) [inline crates/iceberg/src/spec/manifest.rs:1137]: It's better to return `Option<&[u8]>` to make it easier to use.

File: crates/iceberg/src/expr/visitors/expression_evaluator.rs [hunk 2]
```diff
@@ -338,7 +338,7 @@ mod tests {
             nan_value_counts: HashMap::new(),
             lower_bounds: HashMap::new(),
             upper_bounds: HashMap::new(),
-            key_metadata: vec![],
+            key_metadata: Some(vec![]),
```
Comment (Xuanwo) [inline crates/iceberg/src/expr/visitors/expression_evaluator.rs:341]: Do we need to use `None` by default?

**Review by Xuanwo — COMMENTED** [review]
Hi, @feniljain, thanke you for working on this. Here are some minor suggestions.

File: crates/iceberg/src/expr/visitors/expression_evaluator.rs (hunk 2 above)
Comment (feniljain) [reply crates/iceberg/src/expr/visitors/expression_evaluator.rs:341]: They were tests so didn't matter much, but I can change all to `None`

**Review by Xuanwo — COMMENTED** [review]
Thank you, mostly LGTM!

File: crates/iceberg/src/spec/manifest.rs [hunk 3]
```diff
@@ -1164,8 +1164,8 @@ impl DataFile {
         &self.upper_bounds
     }
     /// Get the Implementation-specific key metadata for the data file.
-    pub fn key_metadata(&self) -> &[u8] {
-        &self.key_metadata
+    pub fn key_metadata(&self) -> Option<&[u8]> {
+        self.key_metadata.as_ref().map(|x| x as &[u8])
```
[the remark this answers is not in the document]
Comment (feniljain) [reply crates/iceberg/src/spec/manifest.rs]: Ahh, I was looking for exactly this, thanks for pointing it out! 🙇

**Review by Xuanwo — APPROVED** [review]
Thank you @feniljain for working on this, let's move!

**Pull request merged**